### PR TITLE
Unify tests - all remaining packages

### LINKED ISSUE #249: Unify tests 🧟 👻
:information_source: This is a huge issue that everyone is welcome to work on. Pick a package that was not yet updated and make a pull request.

Do not assign this issue to yourself and do not move it to `In progress` unless you intend to fix all packages.

---

We have decided to unify all tests and use only standard library `testing` package with a small number of test helpers that we write and maintain ourselves.

This task is to gradually (package at a time)
- remove all `testify` package usage and use our [weavetest/assert](https://godoc.org/github.com/iov-one/weave/weavetest/assert) package instead, or just `testing`
- remove all `convey` package usage
- change table tests cases declaration from using slice of structs to [use `map[string]struct{}`](https://github.com/iov-one/weave/blob/master/errors/errors_test.go#L13-L29)

## Packages to be updated
- [ ] `cmd`
- [ ] `cmd/bnsd`   
- [x] `cmd/validators`
- [x] `cmd/lsaddr`
- [x] `cmd/bcpd`
- [ ] `tmtest`
- [x] `scripts`                                                                                                                                                
- [x] `errors`                                                                                                                                                 
- [x] `store`  
- [x] `store/iavl`                                                                                                                                             
- [x] `coin`                                                                                                                                                   
- [x] `orm`
- [x] `app`
- [x] `app/testdata`
- [x] `gconf`
- [x] `crypto`
- [x] `crypto/bech32`
- [x] `x`
- [x] `x/msgfee`
- [x] `x/escrow`
- [x] `x/cash`
- [x] `x/multisig`
- [x] `x/currency`
- [x] `x/batch`
- [x] `x/validators`
- [x] `x/nft`
- [x] `x/hashlock`
- [x] `x/paychan`
- [x] `x/distribution`
- [x] `x/sigs`
- [x] `x/namecoin`
- [x] `x/utils`

### PR DESCRIPTION
**Pull request by alpe**
Resolves #249

### CODE REVIEW

File: x/cash/staticfee_test.go [hunk 1]
```diff
@@ -171,15 +170,15 @@ func TestFees(t *testing.T) {
...
 				err := bucket.Save(kv, wallet)
-				require.NoError(t, err)
+				assert.Nil(t, err)
 			}
 
 			tx := &feeTx{tc.fee}
 
 			_, err := h.Check(nil, kv, tx, &weavetest.Handler{})
-			assert.True(t, tc.expect(err), "%+v", err)
+			assert.Equal(t, true, tc.expect(err))
```
Comment (ruseinov) [inline x/cash/staticfee_test.go]: I suspect this needs to be converted to `IsErr`

File: x/cash/staticfee_test.go [hunk 2]
```diff
@@ -171,15 +170,15 @@ func TestFees(t *testing.T) {
...
 			_, err = h.Deliver(nil, kv, tx, &weavetest.Handler{})
-			assert.True(t, tc.expect(err), "%+v", err)
+			assert.Equal(t, true, tc.expect(err))
```
Comment (ruseinov) [inline x/cash/staticfee_test.go]: Same as above, otherwise we lose the message

File: x/cash/init_test.go [hunk 3]
```diff
@@ -81,16 +81,19 @@ func TestInitState(t *testing.T) {
...
 			if tc.acct != nil {
 				acct, err := bucket.Get(kv, tc.acct)
-				require.NoError(t, err)
-				if assert.NotNil(t, acct) {
-					assert.EqualValues(t, tc.wallet.Coins, AsCoins(acct))
+				assert.Nil(t, err)
+				assert.Equal(t, true, acct != nil)
+				for i = range tc.wallet.Coins {
+					if exp, got := tc.wallet.Coins[i], AsCoins(acct)[i]; !reflect.DeepEqual(exp, got) {
+						t.Errorf("expected %v but got %v", exp, got)
```
Comment (ruseinov) [inline x/cash/init_test.go]: I suspect we can use `assert.Equal(t, exp, got)` here and same some lines.

File: orm/index_test.go [hunk 4]
```diff
@@ -362,14 +369,14 @@ func TestNullableIndex(t *testing.T) {
...
 			for _, init := range tc.setup {
 				err := uniq.Update(db, nil, init)
-				require.NoError(t, err)
+				assert.Nil(t, err)
 			}
 
 			err := uniq.Update(db, tc.prev, tc.next)
 			if tc.isError {
-				require.Error(t, err)
+				assert.Equal(t, true, err != nil)
```
Comment (ruseinov) [inline orm/index_test.go]: I guess we'll need to convert the `tc.isError` here to actual error and use `assert.IsErr`

File: orm/index_test.go [hunk 5]
```diff
@@ -188,33 +188,40 @@ func TestCounterMultiKeyIndex(t *testing.T) {
...
+					assert.Nil(t, idx.insert(db, key, o.Key()))
 				}
 			}
 			// when
 			err := idx.Update(db, spec.prev, spec.next)
 
 			// then
 			if spec.expError {
-				require.Error(t, err)
+				assert.Equal(t, true, err != nil)
```
Comment (ruseinov) [inline orm/index_test.go]: Would be nice to convert and use `isErr`

File: coin/coins_test.go [hunk 6]
```diff
@@ -176,12 +175,12 @@ func TestCombine(t *testing.T) {
 			assert.Equal(t, ac, tc.a.Count())
 			assert.Equal(t, bc, tc.b.Count())
 			if tc.isErr {
-				assert.Error(t, err)
+				assert.Equal(t, true, err != nil)
```
Comment (ruseinov) [inline coin/coins_test.go]: would be nice to convert to use `assert.IsErr`

File: coin/coins_test.go [hunk 7]
```diff
@@ -111,20 +110,20 @@ func TestMakeCoins(t *testing.T) {
 		t.Run(fmt.Sprintf("case-%d", idx), func(t *testing.T) {
 			s, err := CombineCoins(tc.inputs...)
 			if tc.isErr {
-				assert.Error(t, err)
+				assert.Equal(t, true, err != nil)
```
Comment (ruseinov) [inline coin/coins_test.go]: `assert.IsErr` would be preferred here.

File: cmd/bnsd/client/tx_test.go [hunk 8]
```diff
@@ -24,34 +23,34 @@ func TestSendTx(t *testing.T) {
...
+	assert.Nil(t, err)
 	assert.Equal(t, 1, len(conds))
-	assert.EqualValues(t, source.PublicKey().Condition(), conds[0])
+	assert.Equal(t, source.PublicKey().Condition(), conds[0])
 
 	// make sure other chain doesn't validate
 	db = store.MemStore()
 	_, err = sigs.VerifyTxSignatures(db, tx, "foobar")
-	assert.Error(t, err)
+	assert.Equal(t, true, err != nil)
```
Comment (ruseinov) [inline cmd/bnsd/client/tx_test.go]: it would be nice to convert this to `IsErr usage`

File: cmd/bnsd/client/keys_test.go [hunk 9]
```diff
@@ -62,37 +63,37 @@ func TestSaveLoad(t *testing.T) {
...
+	assert.Nil(t, err)
 	loaded, err := LoadPrivateKey(filename)
-	require.NoError(t, err)
+	assert.Nil(t, err)
 	assert.Equal(t, private, loaded)
 
 	// try to over-write, but fails
 	err = SavePrivateKey(private2, filename, false)
-	assert.Error(t, err)
+	assert.Equal(t, true, err != nil)
```
Comment (ruseinov) [inline cmd/bnsd/client/keys_test.go]: `IsErr` would be nice to have here in order to have a proper message.

**Comment by ruseinov** [comment]
Good job! 

I'm going to apologise for the nitpicks now as I know how unrewarding the process of refactoring tests is. 

I have added some comments similar to the reviews we had before on error unification - let's try to keep as much useful info as possible and refactor table tests to be `map[string]interface{}` where it's not the case and use concrete errors to utilise `assert.IsErr` instead of old `tc.isError` booleans. 

All the comments on individual tests for `IsErr` are optional and not a requirement, because these tests are easy to debug anyway. Table tests are harder to debug without having nice `exp, got` error messages.

**Comment by alpe** [comment]
I used search/replace with regExps mostly to replace the testify functions. `IsErr` was not possible with this approach as you would need to know the concrete error. 
I think think this PR adds already value as it fully removes the "testify" dependency from our code. There is always more to improve and polish but let's be pragmatic and tackle not all the problems with a single PR.

> change table tests cases declaration from using slice of structs to use map[string]struct{}

To be fair, I missed that in the original task description. I have added descriptions keys to the range tests I found and were it was possible without much extra work. This should not distract us from removing "testify" though.

**Comment by ruseinov** [comment]
> I used search/replace with regExps mostly to replace the testify functions. `IsErr` was not possible with this approach as you would need to know the concrete error.
> I think think this PR adds already value as it fully removes the "testify" dependency from our code. There is always more to improve and polish but let's be pragmatic and tackle not all the problems with a single PR.
> 
> > change table tests cases declaration from using slice of structs to use map[string]struct{}
> 
> To be fair, I missed that in the original task description. I have added descriptions keys to the range tests I found and were it was possible without much extra work. This should not distract us from removing "testify" though.

Yep, same here. Had to implement it as part of addressing review comments.
While I think it could be cool to remove these booleans same as we did in other PRs I agree that we can merge this asap.

This would probably result in us not touching any of that any time soon, but I don't think this is so important either. 
I will approve and merge this.

**Pull request merged**